### PR TITLE
Refactor: Input data conversion to support conversion messages

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
## Change Summary

Refactor Input data conversion to support conversion messages

## Component(s) name

`arista.avd.validate_and_template`

## Proposed changes

- Move validation and conversion code from action plugin to common lib
- Refactor schema conversion to class
- Return conversion messages for every variable conversion performed

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py]: So here if there are multiple exceptions, only the first one will be raised - is it something we would like to change (even in a future PR and in that case a TODO is enough here)

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py]: Add a docstring

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py]: Probably worth saying what is the format for `path` - a list of strings that give a path in the schema if I get it right?


Missing return type

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py:85]: Exact same code block as from line 87 to 97 - making let's use a helper function

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py:155]: probably worth adding a log message at debug level if this happens

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py:33]: It looks like in this case we expect an AvdSchema object when in other similar classes (validator and schemaresolver) we only need a dict. Lets add the type hint here.

I am not quite sure why we need it to be different as it seems the only place we need this kind of object is if the schema is not given to convert_data we resolve the schema. Would it not be enough to instantiate the class with an already resolved schema?

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py:155]: Added TODOs since this code will be refactored (again) in the deprecation PR. We could add another error type we can yield for this case.

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py]: There is no return type.

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py:85]: Will be done in later PR - the one with deprecation messages.

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py]: added TODO since some of this will change in the later PR.

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py:33]: Added type hint. If we instatiate with resolved schema it just means we need to have more tooling in all the calling functions. We have the same as this in the documentation and schema converts.

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py:33]: Also this would allow for future constants loaded from the schema instance like `ignore keys`

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py]: done

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py:33]: Removed type hint again, since this class is imported into AvdSchema and we cannot have a circular import

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py:33]: I will add a TODO to refactor this to a resolved schema instead. We have several pending PRs in this area, and I think it needs to wait until all those have merged.

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (gmuloc) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py]: technically there is always a return type but ok - https://stackoverflow.com/questions/42227288/pythons-pep-484-type-annotation-for-generator-expression

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py
Comment (gmuloc) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py:33]: resolving but for the future you can have circular import for type hints: https://adamj.eu/tech/2021/05/13/python-type-hints-how-to-fix-circular-imports/

**Review by gmuloc — APPROVED** [review]
LGTM

**Pull request merged**
